hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicyFactory.java

```diff
@@ -0,0 +1,91 @@
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hdds.scm.pipeline;
+
+import org.apache.hadoop.hdds.conf.ConfigurationSource;
+import org.apache.hadoop.hdds.scm.PlacementPolicy;
+import org.apache.hadoop.hdds.scm.ScmConfigKeys;
+import org.apache.hadoop.hdds.scm.container.placement.algorithms.SCMContainerPlacementMetrics;
+import org.apache.hadoop.hdds.scm.container.placement.algorithms.SCMContainerPlacementRackScatter;
+import org.apache.hadoop.hdds.scm.exceptions.SCMException;
+import org.apache.hadoop.hdds.scm.net.NetworkTopology;
+import org.apache.hadoop.hdds.scm.node.NodeManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+
+import static org.apache.hadoop.hdds.scm.ScmConfigKeys.OZONE_SCM_CONTAINER_PLACEMENT_IMPL_KEY;
+
+/**
+ * Pipeline placement factor for pipeline providers to create placement instance
+ * based on configuration property.
+ * {@link ScmConfigKeys#OZONE_SCM_CONTAINER_PLACEMENT_IMPL_KEY}
+ */
+public final class PipelinePlacementPolicyFactory {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(PipelinePlacementPolicyFactory.class);
+
+  private static final Class<? extends PlacementPolicy>
+      OZONE_SCM_CONTAINER_PLACEMENT_IMPL_DEFAULT =
+      SCMContainerPlacementRackScatter.class;
+
+  private PipelinePlacementPolicyFactory() {
+  }
+
+  public static PlacementPolicy getPolicy(
+      ConfigurationSource conf, final NodeManager nodeManager,
+      final PipelineStateManager stateManager,
+      NetworkTopology clusterMap, final boolean fallback,
+      SCMContainerPlacementMetrics metrics) throws SCMException {
+
+    Class<? extends PlacementPolicy> placementClass =
+        PipelinePlacementPolicy.class;
+    Constructor<? extends PlacementPolicy> constructor;
+
+    try {
+      if (conf.get(OZONE_SCM_CONTAINER_PLACEMENT_IMPL_KEY)
+          .equals(SCMContainerPlacementRackScatter.class.getCanonicalName())) {
+        placementClass = conf.getClass(ScmConfigKeys
+                .OZONE_SCM_CONTAINER_PLACEMENT_IMPL_KEY,
+            OZONE_SCM_CONTAINER_PLACEMENT_IMPL_DEFAULT,
+            PlacementPolicy.class);
+
+        constructor = placementClass.getDeclaredConstructor(NodeManager.class,
+            ConfigurationSource.class, NetworkTopology.class, boolean.class,
+            SCMContainerPlacementMetrics.class);
+        LOG.info("Create pipeline placement policy of type {}",
+            placementClass.getCanonicalName());
+        return (constructor.newInstance(nodeManager, conf, clusterMap,
+            fallback, metrics));
+      }
+      constructor = placementClass.getDeclaredConstructor(NodeManager.class,
+          PipelineStateManager.class, ConfigurationSource.class);
+      LOG.info("Create pipeline placement policy of type {}",
+          placementClass.getCanonicalName());
+      return (constructor.newInstance(nodeManager, stateManager, conf));
```

Comment:
We should just use reflection to create a PlacementPolicy and don't make SCMContainerPlacementRackScatter a special case.

Reply:
Currently, the placement polices for pipeline placement differs from those for the replication manager container placement policies.  Because of this, the pipeline policy factory as we have it now, explicitly checks the config for `SCMScaterPlacementRackScatter` set.  If set, then we reuse the `SCMScatterPlacementRackScatter` for pipeline provider pipeline placement as we wanted to do, otherwise the factory uses the existing `PipelinePlacementPolicy` that differs in behavior from container placement policies.

Differing aspects of (pipeline provider) `PipelinePlacementPolicy` to (replication manager) `SCMContainer` placement policies include,
i.) container placement policies do not check for pipeline engagement (we are accommodating this with excluding nodes that have full pipeline engagements in the `RatisPipelineProvider` prior to calling `placementPolicy.chooseDatanodes()`).
ii.) container placement policies enact `SCMContainerPlacementMetrics` that are not collected for pipeline provider placement
iii.)  pipeline provider placement policies, EXCEPT for our implementation support for `RackScatter`, are not configurable.  Prior to our `RackScatter` implementation  the supported pipeline provider pipeline placements,  `Random` and `RackAware`, were chosen within the `PipelinePlacementPolicy` based upon whether the all nodes were on the same rack or not respectively.

Currently, we are unable to directly select the placement policy with reflection for the pipeline provider as we do with the container provider.  Should we continue to explicitly check for SCMContainerPlaementRackScatter here, or is there a way we can use reflection?

Reply:
@neils-dev , please try this https://issues.apache.org/jira/secure/attachment/13053474/4002_review.patch

Reply:
Thanks @szetszwo for the example.  I like it, however having the placement policy set through reflection with first choosing the config `ScmConfigKeys.OZONE_SCM_CONTAINER_PLACEMENT_IMPL_KEY` then fallback to the default, `PipelinePlacementPolicy.class` causes problem should the `ScmConfigKey `be set to something **_other_** than `SCMContainerPlacementRackScatter`.  
For example, the _SCMConfigKey_ can be set to any of the following that are not directly compatible with the pipeline provider pipeline placement policy,
`SCMContainerPlacementCapacity`,
` SCMContainerPlacementRackAware`,
`SCMContainerPlacementRandom`.

Reply:
> ... with first choosing the config ScmConfigKeys.OZONE_SCM_CONTAINER_PLACEMENT_IMPL_KEY ...

We could use a new conf then.  Indeed, we should make it work for all `SCMContainerPlacementCapacity`,
`SCMContainerPlacementRackAware`, `SCMContainerPlacementRandom`.  It could be a follow up work.

Reply:
> We could use a new conf then.

Latest commit reuses the changes made to the `PipelinePlacementFactory` to use reflection to instantiate the `SCMContainerPlacementRackScatter` if set in the new config key `ozone.scm.pipeline.placement.impl`, otherwise to fallback to the default `PipelinePlacementPolicy`.
Made modifications to code affected by this change including the tests.

We can open a follow up Jira to rack re-using the other SCMContainerPlacement policies for the pipeline provider.  We've left the structure (factory and changes to the `RatisPipelineProvider`) used for that in the future here.